File: fixtures/rust-parity/reference_command.py
#!/usr/bin/env python3
"""Reference program used to prove the Rust parity harness contract."""

from __future__ import annotations

import os
from pathlib import Path
import sys


def main() -> int:
    if len(sys.argv) != 3:
        print("usage: reference-command <mode> <root>", file=sys.stderr)
        return 2

    mode = sys.argv[1]
    root = Path(sys.argv[2])
    if mode == "malformed":
        print("malformed input", file=sys.stderr)
        return 65
    if mode == "environment":
        if "FIXTURE_REQUIRED" not in os.environ:
            print("required environment value is unavailable", file=sys.stderr)
            return 69
        return 0
    if mode == "isolation":
        print(f"GH_TOKEN={'present' if 'GH_TOKEN' in os.environ else 'absent'}")
        for key in (
            "GITHUB_API_URL",
            "CLOUDSDK_CONFIG",
            "PATH",
            "LARCH_PARITY_LIVE_SERVICES",
        ):
            print(f"{key}={os.environ[key]}")
        return 0
    if mode != "clean":
        print(f"unknown mode: {mode}", file=sys.stderr)
        return 64

    timestamp = os.environ["FIXTURE_TIMESTAMP"]
    seed = (root / "input" / "seed.txt").read_text(encoding="utf-8").strip()
    output = root / "output"
    output.mkdir()
    rendered = f"root={root}\ntimestamp={timestamp}\nseed={seed}\n"
    (output / "result.txt").write_text(rendered, encoding="utf-8")
    (output / "data.bin").write_bytes(bytes((0, 255)))
    (root / "effects.ndjson").write_text(
        f'{{"action":"write","root":"{root}","at":"{timestamp}"}}\n',
        encoding="utf-8",
    )
    print(f"wrote {root / 'output' / 'result.txt'} at {timestamp}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: fixtures/rust-parity/reference_command.rs
use std::{env, fs, path::Path, process::ExitCode};

fn main() -> ExitCode {
    let arguments: Vec<String> = env::args().collect();
    if arguments.len() != 3 {
        eprintln!("usage: reference-command <mode> <root>");
        return ExitCode::from(2);
    }

    let mode = &arguments[1];
    let root = Path::new(&arguments[2]);
    match mode.as_str() {
        "malformed" => {
            eprintln!("malformed input");
            return ExitCode::from(65);
        }
        "environment" => {
            if env::var_os("FIXTURE_REQUIRED").is_none() {
                eprintln!("required environment value is unavailable");
                return ExitCode::from(69);
            }
            return ExitCode::SUCCESS;
        }
        "isolation" => {
            let token_state = if env::var_os("GH_TOKEN").is_some() {
                "present"
            } else {
                "absent"
            };
            println!("GH_TOKEN={token_state}");
            for key in [
                "GITHUB_API_URL",
                "CLOUDSDK_CONFIG",
                "PATH",
                "LARCH_PARITY_LIVE_SERVICES",
            ] {
                println!(
                    "{key}={}",
                    env::var(key).expect("isolation environment should be set")
                );
            }
            return ExitCode::SUCCESS;
        }
        "clean" => {}
        _ => {
            eprintln!("unknown mode: {mode}");
            return ExitCode::from(64);
        }
    }

    let timestamp = env::var("FIXTURE_TIMESTAMP").expect("fixture timestamp should be set");
    let seed = fs::read_to_string(root.join("input/seed.txt"))
        .expect("seed should be readable")
        .trim()
        .to_owned();
    let output = root.join("output");
    fs::create_dir(&output).expect("output directory should be created");
    let rendered = format!(
        "root={}\ntimestamp={timestamp}\nseed={seed}\n",
        root.display()
    );
    fs::write(output.join("result.txt"), rendered).expect("result should be written");
    fs::write(output.join("data.bin"), [0, 255]).expect("binary result should be written");
    fs::write(
        root.join("effects.ndjson"),
        format!(
            "{{\"action\":\"write\",\"root\":\"{}\",\"at\":\"{timestamp}\"}}\n",
            root.display()
        ),
    )
    .expect("side effect should be recorded");
    println!(
        "wrote {} at {timestamp}",
        output.join("result.txt").display()
    );
    ExitCode::SUCCESS
}
